### ISSUE TITLE
Detect looping over a `map` to compare its key exactly

### ISSUE THREAD

**Issue by gannett-ggreer** [mgechev/revive#1447]
**Is your feature request related to a problem? Please describe.**
In a recent code change, we modified a variable from a slice to a map:
```
-       TagIDs             []string
+       TagIDs             map[string]float64
```
but a loop across the now-map was changed in an obvious but inefficient way:
```
-       for _, id := range a.TagIDs {
+       for id := range a.TagIDs {
                if id == someStaticValue {
                        return nil
                }
        }
```
given that is now a map there's a really direct way to do that test.

**Describe the solution you'd like**
The linter should flag the loop or suggest a fix to make it:
```
if _, exists := a.TagIDs[someStaticValue]; exists {
    return nil
}
```

**Describe alternatives you've considered**
This isn't a correctness issue, just inefficiency, so when in doubt it shouldn't be too picky.

**Additional context**
This happens remarkably often in my experience for people new to Go or who are making scattered changes and don't always think too deeply about fixing the resulting compilation errors as they focus on the end goal.  None of the `golangci-lint` linters or current version of `revive` tripped on that loop as anything of note, but it is also more of a learning opportunity for the programmer rather than anything wrong necessarily with the program.

**Comment by alexandear** [comment]
Thanks for the suggestion!

While the idea makes sense from an efficiency and readability perspective, this type of pattern—looping over a map to check for the presence of a single key—is more of a performance/style concern than a correctness issue. Introducing a rule for this case may lead to false positives or overly opinionated suggestions.


That said, this could be a great candidate for a more targeted static analysis tool or as a custom rule in projects where this pattern is a common pain point.

@ccoVeille @chavacava what do you think?

**Comment by gannett-ggreer** [comment]
It is similar in a sense to linters that say "you're not using any %-formatting codes in `fmt.Errorf`, you should use `errors.New` instead".

For example:
```
authoring/embed.go:40:11: use-errors-new: replace fmt.Errorf by errors.New (revive)
			Err:   fmt.Errorf("no body found for embed"),
```

**Comment by ccoVeille** [comment]
I'm fine with having the rule described in the issue in revive

**Comment by chavacava** [comment]
`revive` already has (too many?) "style-related" (enforce-*, max-*, ...) rules, one more or one less will not change affairs so why not.

The rule will need type information (to recognise loops over maps)

**Comment by chavacava** [comment]
The branch [rule/inefficient-map-lookup](https://github.com/mgechev/revive/tree/rule/inefficient-map-lookup) implements a draft of the rule `inefficient-map-lookup` that aims to handle the use case described in this issue.

The rule looks for ranges on map keys where the body is just one if on the key. If the range body has more than one statement the rule will not check it.

Some findings of the rule on public repositories:

https://github.com/open-policy-agent/opa/blob/f77322b3fb8ba5555adc6493e120a205dd2860a7/v1/bundle/keys.go#L54-L59

https://github.com/golang/tools/blob/6ebfec2837bbfbbdc5d3a9fe087c70c29be27fbf/go/packages/golist.go#L315-L321

@gannett-ggreer could you please check if the rule draft copes with your case?

**Comment by gannett-ggreer** [comment]
Yes, it works for the original code case I mentioned.
